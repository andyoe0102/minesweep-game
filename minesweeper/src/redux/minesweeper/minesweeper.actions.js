export const resetGame = () =>({
    type:'RESET_GAME',

});

export const addHeight = ()=>({
    type:'addHeight'
})

export const checkMine = (cell)=>({
    type:'CHECK_MINE',
    payload:cell
})


export const flagCell = (cell)=>({
    type:'FLAG_CELL',
    payload:cell
})
