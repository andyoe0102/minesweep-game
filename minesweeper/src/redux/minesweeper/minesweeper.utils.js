function sampleSize(array, n) {
    n = n == null ? 1 : n
    let length = array == null ? 0 : array.length
    if (!length || n < 1) {
      return []
    }
    n = n > length ? length : n
    let index = -1
    let lastIndex = length - 1
    let result = [...array];
    while (++index < n) {
      let rand = index + Math.floor(Math.random() * (lastIndex - index + 1))
      let value = result[rand]
      result[rand] = result[index]
      result[index] = value
    }
    return result.slice(0,n)
  }

function defineCells(state,mines){
    return{
        ...state,
        grid: state.grid.map((row)=>{
            return row.map((col)=>{
                return mines.includes(col.id)? {...col,mine:true} : col
            })
        })
    }
}

function checkNeighbor(state,x,y){
    if(x < 0 || y < 0 || x>9 || y>9){
        return 0
    };

    if(state.grid[y][x].mine===true){
        return 1
    }else{
        return 0
    }
}

function countNeighbors(state,cell){
    const {x,y} = cell;
    if(cell.mine===true){
        return ''
    }

    let topleft = checkNeighbor(state,x-1,y-1);
    let top = checkNeighbor(state,x,y-1);
    let topright = checkNeighbor(state,x+1,y-1);
    let left = checkNeighbor(state,x-1,y);
    let right = checkNeighbor(state,x+1,y);
    let bottomleft = checkNeighbor(state,x-1,y+1);
    let bottom= checkNeighbor(state,x,y+1);
    let bottomright=checkNeighbor(state,x+1,y+1);
    
    let total = topleft + top + topright+left+right+bottomleft+bottom+bottomright

    return total;
}

const findNeighbors = (state,cell) =>{
    const {x,y} = cell;

    function checkExist(state,x,y){
        if(x<0 || y <0 | x>9 || y> 9){
            return undefined
        }

        if(state.grid[y][x] ===undefined){
            return undefined
        }else{
            return state.grid[y][x]
        }
    }

    let array = [
        checkExist(state,x-1,y-1),
        checkExist(state,x,y-1),
        checkExist(state,x+1,y-1),
        checkExist(state,x-1,y),
        checkExist(state,x+1,y),
        checkExist(state,x-1,y+1),
        checkExist(state,x,y+1),
        checkExist(state,x+1,y+1)
    ]

    return array.filter((idx)=>{
        return idx !==undefined
    })
}

export const placemines = (state,cellId) => {
    let cells = state.grid.flat().filter((cell)=>cell.id !== cellId).map((cell)=>cell.id);

    let mines = sampleSize(cells,state.mineCount);

    let newCells = defineCells(state,mines);

    return{
        ...defineCells(state,mines),
        started:true
    }
}
    


export const addneighbors= (state)=>{
    let total = 0;
    return {    
        ...state,
        grid: state.grid.map((row)=>{
            return row.map((col)=>{
                let neighborCount = countNeighbors(state,col);
                return {...col,neighborCount:neighborCount}
        })
    })
    }
    
}

export const checkWin = (state) =>{
    let totalRevealed = 0;
    state.grid.map((row)=>{
        return row.map((col)=>{
            if(col.revealed){
                totalRevealed+=1
            }
        })
    })

    if(totalRevealed === 90){
        return true
    }else{
        return false
    }
}


export const checkMines = (state,cell,checked=[])=>{
    let NeighborCells = findNeighbors(state,cell);

    if(cell.mine){
        return [cell]
    }

    checked.push(cell)
    if(cell.neighborCount >0){
        return [cell]
    }else{
        let ids = checked.map((obj)=>{return obj.id})
        let checkMinesArray = NeighborCells.filter((neighborcell)=>{

            return neighborcell.flagged === false && ids.includes(neighborcell.id) === false;
        })


        return [cell,
            ...checkMinesArray.map((neighborcell)=> checkMines(state,neighborcell,checked))
        ].flat()
    }


}

export const removeDuplicates=(originalArray, prop)=> {
    var newArray = [];
    var lookupObject  = {};

    for(var i in originalArray) {
       lookupObject[originalArray[i][prop]] = originalArray[i];
    }

    for(i in lookupObject) {
        newArray.push(lookupObject[i]);
    }
     return newArray;
}

export const revealCells =(state,array)=>{
    let ids = array.map((cell)=>{
        return cell.id
    });

    return {
        ...state,
        grid:state.grid.map((row)=>{
            return row.map((col)=>{
                return ids.includes(col.id)? ({...col, revealed:true}):col
            })
        })
    }
}   

export const flagCell = (state,cell)=>{

    if(cell.revealed){
        return state
    }else{
        let newstate ={
            ...state,
            grid: state.grid.map((row)=>{
                return row.map((col)=>{
                    return cell.id === col.id?({...col, flagged: !col.flagged}): col
                })
            })
        }

        return newstate
    }
    return state
}

