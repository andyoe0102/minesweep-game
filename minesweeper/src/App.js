import React from 'react';
import {connect} from 'react-redux';

import Grid from './components/Grid';

import {resetGame,addHeight,checkMine} from './redux/minesweeper/minesweeper.actions';

import './App.css';

class App extends React.Component{
  constructor(props){
    super(props);


  }

  render(){
    return (
      <div className="App">
        <Grid></Grid>
        <button className="resetButton" onClick={this.props.reset}>ResetGame</button>

      </div>
    );
  }
  
}

const mapDispatchToProps = dispatch => ({
  reset: () => dispatch(resetGame()),
  checkMine: (cellId) => dispatch(checkMine(cellId))
})

const mapStateToProps = state => ({
  minesweeper: state.minesweeper
})

export default connect(mapStateToProps,mapDispatchToProps)(App);
