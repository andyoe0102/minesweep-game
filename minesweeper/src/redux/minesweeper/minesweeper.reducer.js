import {placemines,addneighbors,checkWin,checkMines,removeDuplicates,revealCells,flagCell} from './minesweeper.utils'

const INITIAL_STATE ={
    width : 10,
    height : 10,
    mineCount : 10,
    started:false,
    win:false,
    lose:false,
    grid: Array(10).fill(0).map((idx,key)=>{
        return Array(10).fill(0).map((idx2,key2)=>{
            return {
                id: `${key2}-${key}`,
                x: key2,
                y:key,
                flagged:false,
                mine:false,
                revealed:false,
                neighborCount:''
            }
        })
    })
}

const minesweepReducer = (state = INITIAL_STATE, action)=>{
    switch(action.type){
        case 'RESET_GAME':
            return INITIAL_STATE;

        case 'CHECK_MINE':
            if(state.win || state.lose){
                return state
            }

            if(!state.started){
                console.log('first search!')
                state = placemines(state,action.payload.id);
                let newstate = addneighbors(state);

                let {x,y} = action.payload
                let cell = newstate.grid[y][x]

                let checkedMines = checkMines(newstate,cell);
                let uniqueArray = removeDuplicates(checkedMines, "id");
    
                let revealedState = revealCells(newstate,uniqueArray)



                return {
                    ...revealedState
                }
            }else{
                console.log('not first search!')

                if(action.payload.revealed){
                    console.log('already relvealed!')
                    return state;
                }
                let checkedMines = checkMines(state,action.payload);
                let uniqueArray = removeDuplicates(checkedMines, "id");
  
                let revealedState = revealCells(state,uniqueArray)
                let lose = action.payload.mine;
                let win = checkWin(revealedState);

                return {
                    ...revealedState,
                    win,
                    lose
                }

            }
        case 'FLAG_CELL':

            if(!state.started){


                state = placemines(state,action.payload.id);
                let newstate = addneighbors(state);

                let {x,y} = action.payload
                let cell = newstate.grid[y][x]

                newstate = flagCell(newstate,cell)

                return newstate;
            }


            let newstate = flagCell(state,action.payload)


            return newstate



            
            

        default:
            return state;
    }
}

export default minesweepReducer;

