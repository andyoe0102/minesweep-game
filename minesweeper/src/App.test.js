import React from 'react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import App from './App';
import * as actions from './redux/minesweeper/minesweeper.actions'

 
const mockStore = configureStore([]);
 
describe('My Connected React-Redux Component', () => {
  let store;
  let component;
  const INITIAL_STATE ={
    width : 10,
    height : 10,
    mineCount : 10,
    started:false,
    win:false,
    lose:false,
    grid: Array(10).fill(0).map((idx,key)=>{
        return Array(10).fill(0).map((idx2,key2)=>{
            return {
                id: `${key2}-${key}`,
                x: key2,
                y:key,
                flagged:false,
                mine:false,
                revealed:false,
                neighborCount:''
            }
        })
    })
}
 
  beforeEach(() => {
    store = mockStore({
      minesweeper: INITIAL_STATE,
    });

    store.dispatch = jest.fn();
 
    component = renderer.create(
      <Provider store={store}>
        <App />
      </Provider>
    );
  });
 
  it('should render with given state from Redux store', () => {
    expect(component.toJSON()).toMatchSnapshot();
  });
 
  it('should dispatch an action on button click', () => {
    renderer.act(() => {
      component.root.findByType('button').props.onClick();
    });
 
    expect(store.dispatch).toHaveBeenCalledTimes(1);
    
    expect(store.dispatch).toHaveBeenCalledWith(
      actions.resetGame({ type:'RESET_BUTTON' })
    );
  });
});


