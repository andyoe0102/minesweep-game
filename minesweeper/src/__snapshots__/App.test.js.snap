// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`My Connected React-Redux Component should render with given state from Redux store 1`] = `
<div
  className="App"
>
  <div
    className="grid"
  >
    <div
      className="row"
    >
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
      <div
        className="cell"
      >
        <img
          alt="pic"
          className="minesweeperimg"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          src="board_blank_unclicked.png"
        />
      </div>
    </div>
  </div>
  <button
    className="resetButton"
    onClick={[Function]}
  >
    ResetGame
  </button>
</div>
`;
